src/main/java/com/arpnetworking/utility/RegexAndMapReplacer.java

```diff
@@ -0,0 +1,216 @@
+            // Append left-over string after the matches
+            if (lastMatchedIndex < input.length() - 1) {
+                builder.append(input.substring(lastMatchedIndex, input.length()));
+            }
+            return new Replacement(builder.toString(), variablesUsedBuilder.build());
+        }
+        return new Replacement(input, ImmutableList.of());
+    }
+
+    private static void appendReplacement(
+            final Matcher matcher,
+            final String replacement,
+            final StringBuilder replacementBuilder,
+            final Map<String, String> variables,
+            final ImmutableList.Builder<String> variablesUsedBuilder) {
+        final StringBuilder tokenBuilder = new StringBuilder();
+        int x = -1;
+        while (x < replacement.length() - 1) {
+            x++;
+            final char c = replacement.charAt(x);
+            if (c == '\\') {
+                x++;
+                processEscapedCharacter(replacement, x, replacementBuilder);
+            } else {
+                if (c == '$') {
+                    x += writeReplacementToken(replacement, x, replacementBuilder, matcher, variables, tokenBuilder, variablesUsedBuilder);
+                } else {
+                    replacementBuilder.append(c);
+                }
+            }
+        }
+    }
+
+    private static void processEscapedCharacter(final String replacement, final int x, final StringBuilder builder) {
+        if (x >= replacement.length()) {
+            throw new IllegalArgumentException(
+                    String.format("Improper escaping in replacement, must not have trailing '\\' at col %d: %s", x, replacement));
+        }
+        final Character c = replacement.charAt(x);
+        if (c == '\\' || c == '$' || c == '{' || c == '}') {
+            builder.append(c);
+        } else {
+            throw new IllegalArgumentException(
+                    String.format("Improperly escaped '%s' in replacement at col %d: %s", c, x, replacement));
+        }
+    }
+
+    private static int writeReplacementToken(
+            final String replacement,
+            final int offset,
+            final StringBuilder output,
+            final Matcher matcher,
+            final Map<String, String> variables,
+            final StringBuilder tokenBuilder,
+            final ImmutableList.Builder<String> variablesUsedBuilder) {
+        boolean inReplaceBrackets = false;
+        boolean tokenNumeric = true;
+        tokenBuilder.setLength(0);  // reset the shared builder
+        int x = offset + 1;
+        char c = replacement.charAt(x);
+
+        // Optionally consume the opening brace
+        if (c == '{') {
+            inReplaceBrackets = true;
+            x++;
+            c = replacement.charAt(x);
+        }
+
+        if (inReplaceBrackets) {
+            // Consume until we hit the }
+            while (x < replacement.length() - 1 && c != '}') {
+                if (c == '\\') {
```

Comment:
Like regex capture groups, there should be no need to support escaped characters inside replacements unless you want to support dynamic replacements; 

e.g.
`${foo_${bar}}`

This would certainly give insane flexibility.

Reply:
Allowing escaping does not allow for dynamic replacements like that.  It does allow for "special" characters in the names of the dimension, however. For example, if someone decided to have "{a}" as a dimension name, for whatever reason.  Moreover, it makes for more consistent rules about how to escape (especially important if you have multiple levels of escaping.

Reply:
Added the prefix support (env, capture, dimension are supported now).  Precedence is explicit in the ordering of the prefixed variable maps.  Phew, that was a bit of work.